### ISSUE TITLE
[Feature] Allow tracing

### ISSUE THREAD

**Issue by ruifigueira** [ruifigueira/playwright-crx#7]
With `memfs` we can get tracings. It captures screenshots but no snapshots.

Here is an example (it opens the memfs file in trace.playwright.dev):

```ts
test('should trace', async ({ runCrxTest, _extensionServiceWorkerDevtools }) => {
  await runCrxTest(async ({ crxApp, page, context }) => {
    await context.tracing.start({ snapshots: true, screenshots: true, sources: true });
    await page.goto('https://demo.playwright.dev/todomvc/#/');
    await page.getByPlaceholder('What needs to be done?').fill('Playwright CRX rocks!');
    await page.getByPlaceholder('What needs to be done?').press('Enter');
    await page.getByLabel('Toggle Todo').click();
    await page.getByRole('button', { name: 'Clear completed' }).click();
    await context.tracing.stop({ path: '/crx/trace.zip' });

    const tracePage = await crxApp.newPage();
    await tracePage.goto('https://trace.playwright.dev');

    const [chooser] = await Promise.all([
      tracePage.waitForEvent('filechooser'),
      tracePage.getByRole('button', { name: 'Select file(s)' }).click(),
    ]);
    await chooser.setFiles('/crx/trace.zip');
    debugger;
  });
});
```

Problem seems to be related with the way [playwright determines the API name](https://github.com/microsoft/playwright/blob/1aee48f2d065cc899e862dedef536ddda382f618/packages/playwright-core/src/utils/stackTrace.ts#L75-L85).

Basically, it uses the stacktrace but because we're using vite, which compiles to a single file, it's not able to determine the api name and, without it, it doesn't trigger the instrumentation events.

**Comment by jasonwcfan** [comment]
Hi @ruifigueira did you manage to get tracing to work in the newest version of playwright-crx? I tried cherry picking [this commit](https://github.com/ruifigueira/playwright-crx/commit/6644cad9acaa21d7a1ac2ddc9d5a6444d37d9f0e) into a fork but the `should trace` test fails when trying to upload `/crx/trace.zip` to trace.playwright.dev.

It works when building from that commit, but not after merging with the most recent changes. I suspect 

Awesome project btw! Looking to use it in a web automation project I'm working on.

**Comment by ruifigueira** [comment]
Hi @jasonwcfan , I didn't merge that at the time because it didn't work ootb, it required the final bundle to be explicitly registered in `_stacktraceSourcemap`, which was far from ideal.

I'll take a second look into it to see if I come up with a better solution.

**Comment by jasonwcfan** [comment]
Thanks for the reply @ruifigueira . I'm looking into it as well, if I get it working I'll put up a PR

**Comment by ruifigueira** [comment]
I did the same here, same results as you. Looking at `tests/test-extension/dist/background.js.map`, I would bet the issue comes from there, as the file is basically filled with semicolons:

![image](https://github.com/user-attachments/assets/c01baf1d-c55b-4cf4-b12a-7d9f69b9dd88)

Can you check on your side if you're getting the same?

**Comment by ruifigueira** [comment]
I suspect that maybe `rollup-plugin-sourcemaps` doesn't go well with vite 5.
I was able to make it work by keeping playwright as external library (for now, I'm copying it directly to `test-extension/dist`):

![image](https://github.com/user-attachments/assets/0ce88764-40a9-4af1-b77e-7c323c71f9e6)

I'll try to set it up properly, because right now I'm doing some manual changes directly on the dist folder :)

**Comment by ruifigueira** [comment]
I didn't spend too much time on it, there's lots of nasty code, but the idea is there: dd68fb7.

**Comment by jasonwcfan** [comment]
Amazing, thanks @ruifigueira. 

I tried running it but encountered a build issue:
```
> playwright-crx-tests@0.10.0 build
> tsc -p ./test-extension/tsconfig.json && vite build ./test-extension

test-extension/src/background.ts:17:31 - error TS2305: Module '"playwright-crx/test"' has no exported member '_fs'.

17 import { crx, expect, _debug, _fs } from 'playwright-crx/test';
                                 ~~~


Found 1 error in test-extension/src/background.ts:17
```

I thought that Vite would automatically be replacing references to `playwright-crx/test` with `./playwright-crx/test.mjs` per the vite config for `test-extension`, but that doesn't seem to be happening? I'm not very familiar with Vite so I'm not sure what's going on, maybe it's an issue with the build order.

**Comment by ruifigueira** [comment]
vite by default tries to bundle everything (similar to static linking), so test-extension is loading it as an external library. But for that, the dist folder must contain those files. In theory, this code should be doing that, as well as replacing the import in the compiled `background.js`:

https://github.com/ruifigueira/playwright-crx/blob/dd68fb7cceff53c3b6728a44f5984c5df956aa6b/tests/test-extension/vite.config.ts#L25-L32

But maybe it's not working for you.

Try this:

```
npm run build:crx
npm run build:tests
````

then copy `lib/test.mjs` and `lib/test.mjs.map` into `tests/test-extension/dist`, and edit `tests/test-extension/dist/background.js` to import `./playwright-crx/test.mjs`:

```
import { crx, expect, _debug, _fs } from "./playwright-crx/test.mjs";
```

 before running the test:

```
npm run test -- --grep 'should trace'
```

**Comment by jasonwcfan** [comment]
@ruifigueira Yeah it's weird, but I got it working by just adding `// @ts-ignore` above the import. There was also one line in vite.config.ts that needed to be uncommented out (line 112).

I got the test working initially, but for some reason it's not working anymore. It's unrelated to tracing, all tests in `api.spec.ts` are failing (even the smoketest). Other tests are running fine.

```
 1) [Chrome] › crx/api.spec.ts:21:5 › should work @smoke ──────────────────────────────────────────

    Error: worker.evaluate: Target page, context or browser has been closed

       at crx/crxTest.ts:105

      103 |
      104 |     // wait for initialization
    > 105 |     await worker.evaluate(() => new Promise<void>((resolve, reject) => {
          |                  ^
      106 |       if (serviceWorker.state !== 'activated') {
      107 |         serviceWorker.addEventListener('statechange', () => {
      108 |           if (serviceWorker.state === 'activated') resolve();

        at Object.extensionServiceWorker (/Users/jasonfan/Documents/code/playwright-crx/tests/crx/crxTest.ts:105:18)
```

It must be an issue with some configuration somewhere as they're working for the previous trace commit. No need to look into it if you're busy as you've already been extremely helpful, but in case you've seen this before and know how to solve it let me know!

**Comment by ruifigueira** [comment]
That error normally occurs when the test extension (`tests/test-extension/dist`) doesn't load properly. Try to load it in a chrome and check the extension devtools console, to see if there are errors there.

**Comment by ruifigueira** [comment]
I updated https://github.com/ruifigueira/playwright-crx/tree/feat/tracing, with a latest release of vite the issues with source map disapeared. I also simplified it a lot, hopefully I'll include this in 0.12.0, in sync with playwright 1.50.0.

**Comment by ruifigueira** [comment]
I think it would be possible but it's not trivial and I need to investigate a little bit more. Playwright's current implementation processes the stacktrace to extract the JS file path and from there infer if it's a API call or not.

Of course, we know all API functions, and we could inject some javascript code on each of them (for instance, using vite) that registers that call.

**Comment by ruifigueira** [comment]
Well, you can do that low level right now, by using [server-side tracing instance](https://github.com/ruifigueira/playwright-crx/blob/main/playwright/packages/playwright-core/src/server/trace/recorder/tracing.ts), something like:

```js
    await page.context().tracing.start();
    // tracingImpl is the server-sid	e tracing instance
    const tracingImpl = await (page.context().tracing as any)._toImpl();
    await tracingImpl.onBeforeInputAction((page as any)._toImpl(), {
      id: '',
      apiName: 'locator.click',
      startTime: 0,
      endTime: 0,
      method: '',
      params: {},
      log: [],
      isServerSide: true,
    });
    await page.context().tracing.stop();
```

**Comment by ruifigueira** [comment]
@ducan-ne in case you're interested, I'm removing the requirement of sourcemaps for tracing. You can take a look at #54

**Comment by ruifigueira** [comment]
> - crxApp.close() is still unresponsive (calling .close() remains pending)

I'll investigate it.

> - I tried your example code but it seems like doesn't work, not sure because of me or something

You were right, you first need to call `tracingImpl.onBeforeCall`, just like this:

```ts
test('should add event', async ({ runCrxTestAndParseTraceRaw }) => {
  const trace = await runCrxTestAndParseTraceRaw(async ({ context, page }) => {
    const callMetadata = {
      id: 'call@123',
      apiName: 'locator.click',
      startTime: Date.now(),
      endTime: Date.now(),
      method: 'locator.click',
      params: {},
      log: [],
    };
    await context.tracing.start({ screenshots: true, snapshots: true });
    const pageImpl = (page as any)._toImpl();
    const tracingImpl = await (page.context().tracing as any)._toImpl();
    await tracingImpl.onBeforeCall(pageImpl, callMetadata);
    await tracingImpl.onBeforeInputAction(pageImpl, callMetadata);
    await context.tracing.stopChunk({ path: 'trace.zip' });
  });

  expect(trace.actions).toEqual([
    'locator.click',
  ]);
});
```

**Comment by ruifigueira** [comment]
Keep in mind that a chrome extension has not access to actual filesystem. I use memfs to have a memory implementation of node's fs API for everything to work.

To get the actual trace, you either have to upload it somewhere (in README you have an example where I upload it to trace.playwright.dev) or download it.

**Comment by ruifigueira** [comment]
if you set the absolute path:

```
await context.tracing.stopChunk({ path: '/tmp/trace.zip' });
```

It should work.

**Comment by ruifigueira** [comment]
Can you paste the code you're testing? On my side, it's working properly:

![image](https://github.com/user-attachments/assets/62e353c1-3bdf-45b5-98f9-cdfc1ec0463c)

I'm testing with the following code:

```ts
test('should create trace file', async ({ runCrxTest }) => {
  await runCrxTest(async ({ context, page, server, fs, expect }) => {
    await page.goto(server.PREFIX + '/input/button.html');
    await context.tracing.start({ screenshots: true, snapshots: true });
    await page.click('button');
    await context.tracing.stop({ path: '/tmp/trace.zip' });
    expect(fs.existsSync('/tmp/trace.zip')).toBe(true);
    console.log(fs.readdirSync('/tmp'));
    debugger;
  });
});
```

**Comment by ruifigueira** [comment]
I just tried your code and it's working:

![image](https://github.com/user-attachments/assets/6ad82837-a007-446e-9f83-1dec74753bd5)

I just replaced todomvc code with yours.

Are you importing memfs on your own? In this comment of yours I suspect you're not using `crx.fs` and instead you're using another `memfs` instance (just like using a different filesystem).

Try to inspect `crx.fs.readdirSync('/crx')` instead.

**Comment by ruifigueira** [comment]
If you could provide a code example that reproduces the error, it would be great.

> Secondly, crxApp.close() works, but the "started debugging this browser" banner doesn't disappear immediately, it takes a few seconds

That's normal, detaching `chrome.debugger` API works like that, it takes a few seconds for the banner to disappear. I assume it's to avoid that bad unintended extensions can attach and detach the debugger API quickly without the user having time to notice it.

**Comment by ruifigueira** [comment]
I was able to replicate it, and even in a test:

```js
test('should be able to open new crxApp after close', async ({ runCrxTest }) => {
  await runCrxTest(async ({ crx, crxApp, expect }) => {
    await crxApp.close();
    expect(await crx.get()).toBeUndefined();
    const newCrxApp = await crx.start();
    expect(await crx.get()).toBe(newCrxApp);
    // ensure we can open a new page on the newly opened crxApp
    // see: https://github.com/ruifigueira/playwright-crx/issues/7#issuecomment-2570047648
    const newPage = await newCrxApp.newPage();
    expect(newPage).toBeTruthy();
  });
});
```

**Comment by ruifigueira** [comment]
Fixed in cf69af5, already in main branch
